{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> GTF File Format Description : <a href=\"https://www.gencodegenes.org/data_format.html\"> Gencode </a></h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes Stats:\n",
      "\n",
      "1-unitary_pseudogene: 95\n",
      "2-rRNA: 544\n",
      "3-lincRNA: 7490\n",
      "4-IG_C_pseudogene: 9\n",
      "5-translated_processed_pseudogene: 2\n",
      "6-Mt_tRNA: 22\n",
      "7-sense_intronic: 899\n",
      "8-IG_V_gene: 144\n",
      "9-misc_RNA: 2213\n",
      "10-pseudogene: 18\n",
      "11-Mt_rRNA: 2\n",
      "12-polymorphic_pseudogene: 38\n",
      "13-scaRNA: 49\n",
      "14-IG_J_gene: 18\n",
      "15-TR_J_pseudogene: 4\n",
      "16-IG_J_pseudogene: 3\n",
      "17-scRNA: 1\n",
      "18-TEC: 1067\n",
      "19-protein_coding: 19901\n",
      "20-bidirectional_promoter_lncRNA: 47\n",
      "21-TR_V_pseudogene: 33\n",
      "22-sense_overlapping: 183\n",
      "23-TR_J_gene: 79\n",
      "24-TR_D_gene: 4\n",
      "25-IG_V_pseudogene: 188\n",
      "26-antisense: 5501\n",
      "27-snRNA: 1900\n",
      "28-sRNA: 5\n",
      "29-unprocessed_pseudogene: 2664\n",
      "30-TR_V_gene: 106\n",
      "31-transcribed_unprocessed_pseudogene: 854\n",
      "32-3prime_overlapping_ncRNA: 32\n",
      "33-miRNA: 1881\n",
      "34-non_coding: 3\n",
      "35-macro_lncRNA: 1\n",
      "36-IG_C_gene: 14\n",
      "37-ribozyme: 8\n",
      "38-IG_D_gene: 37\n",
      "39-TR_C_gene: 6\n",
      "40-IG_pseudogene: 1\n",
      "41-processed_transcript: 556\n",
      "42-transcribed_processed_pseudogene: 472\n",
      "43-processed_pseudogene: 10219\n",
      "44-snoRNA: 943\n",
      "45-vaultRNA: 1\n",
      "46-transcribed_unitary_pseudogene: 123\n",
      "________________________________________________________________________________\n",
      "Total: 58380\n",
      "\n",
      "\n",
      "Transcripts Stats: \n",
      "\n",
      "1-unitary_pseudogene: 95\n",
      "2-rRNA: 544\n",
      "3-lincRNA: 13598\n",
      "4-IG_C_pseudogene: 9\n",
      "5-translated_processed_pseudogene: 2\n",
      "6-retained_intron: 27834\n",
      "7-sense_intronic: 958\n",
      "8-IG_V_gene: 144\n",
      "9-polymorphic_pseudogene: 59\n",
      "10-Mt_rRNA: 2\n",
      "11-misc_RNA: 2227\n",
      "12-Mt_tRNA: 22\n",
      "13-scaRNA: 49\n",
      "14-IG_J_gene: 18\n",
      "15-TR_J_pseudogene: 4\n",
      "16-IG_J_pseudogene: 3\n",
      "17-scRNA: 1\n",
      "18-TEC: 1163\n",
      "19-protein_coding: 82335\n",
      "20-bidirectional_promoter_lncRNA: 243\n",
      "21-TR_V_pseudogene: 33\n",
      "22-sense_overlapping: 350\n",
      "23-TR_J_gene: 79\n",
      "24-TR_D_gene: 4\n",
      "25-IG_V_pseudogene: 188\n",
      "26-nonsense_mediated_decay: 14889\n",
      "27-pseudogene: 37\n",
      "28-snRNA: 1900\n",
      "29-sRNA: 5\n",
      "30-unprocessed_pseudogene: 2665\n",
      "31-TR_V_gene: 106\n",
      "32-transcribed_unprocessed_pseudogene: 862\n",
      "33-3prime_overlapping_ncRNA: 40\n",
      "34-miRNA: 1881\n",
      "35-non_coding: 3\n",
      "36-macro_lncRNA: 1\n",
      "37-antisense: 11161\n",
      "38-IG_C_gene: 23\n",
      "39-ribozyme: 8\n",
      "40-IG_D_gene: 37\n",
      "41-TR_C_gene: 6\n",
      "42-vaultRNA: 1\n",
      "43-IG_pseudogene: 1\n",
      "44-processed_transcript: 28382\n",
      "45-transcribed_processed_pseudogene: 472\n",
      "46-processed_pseudogene: 10221\n",
      "47-snoRNA: 955\n",
      "48-non_stop_decay: 87\n",
      "49-transcribed_unitary_pseudogene: 128\n",
      "________________________________________________________________________________\n",
      "Total: 203835\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def get_stats(types):\n",
    "    count = 1\n",
    "    total = 0\n",
    "    for key,val in types.iteritems():\n",
    "        total += val\n",
    "        print \"%d-%s: %d\" % (count, key, val)\n",
    "        count += 1\n",
    "        \n",
    "    print \"_\" * 80\n",
    "    print \"Total: %s\\n\\n\" % (total)\n",
    "\n",
    "gene_types = {}\n",
    "transcript_types = {}\n",
    "    \n",
    "with open(\"gencode.v28.annotation.gtf\",\"r\") as gtf:\n",
    "    next(gtf);next(gtf);next(gtf);next(gtf);next(gtf);next(gtf);\n",
    "    for line in gtf:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        feature_type = fields[2]\n",
    "        info = fields[-1].split(\"; \")\n",
    "\n",
    "        if feature_type == \"gene\":\n",
    "            gene_type = info[1].replace('\"','').split()[1]\n",
    "            if gene_type in gene_types:\n",
    "                gene_types[gene_type] += 1\n",
    "            else:\n",
    "                gene_types[gene_type] = 1\n",
    "        \n",
    "        elif feature_type == \"transcript\":\n",
    "            transcript_type = info[4].replace('\"','').split()[1]\n",
    "            if transcript_type in transcript_types:\n",
    "                transcript_types[transcript_type] += 1\n",
    "            else:\n",
    "                transcript_types[transcript_type] = 1\n",
    "            \n",
    "        else:\n",
    "            continue\n",
    "\n",
    "\n",
    "print \"Genes Stats:\\n\"\n",
    "get_stats(gene_types)\n",
    "print \"Transcripts Stats: \\n\"\n",
    "get_stats(transcript_types)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
